Use RandomAccess/File.OpenHandle in a few more places

stephentoub opened the pull request:
Places where we don't need to pay for the FileStream and the access pattern is easily replaced.

stephentoub commented:
This fails because of MS_IO_REDIST, and it's too annoying to use ifdefs everywhere for it.  We should really consider stopping building this package.

stephentoub commented:
Updated with undos for the File changes that would have required large ifdefs.

adamsitnik commented:
> This fails because of MS_IO_REDIST, and it's too annoying to use ifdefs everywhere for it. We should really consider stopping building this package.

We plan to reevaluate this package for .NET 7 and hopefully remove it https://github.com/dotnet/runtime/issues/54919#issuecomment-870988742 cc @jeffhandley

adamsitnik commented on src/libraries/System.Private.CoreLib/src/System/IO/File.cs:
```diff
@@ -373,19 +374,13 @@ public static void WriteAllBytes(string path, byte[] bytes)
+            fs.Write(bytes, 0, bytes.Length);
+#else
+            using SafeFileHandle sfh = OpenHandle(path, FileMode.Create, FileAccess.Write, FileShare.Read);
+            RandomAccess.Write(sfh, bytes, 0);
```
I would love to use `RandomAccess` here and even in more places, however, I have one concern: what about paths that point to pipes or other unseekable files?

On Windows we don't allow for opening `FileStream` from a path that points to a non-regular file:

https://github.com/dotnet/runtime/blob/0696727ece4d363f171f80dd442b61fb20c6500a/src/libraries/System.Private.CoreLib/src/System/IO/Strategies/FileStreamHelpers.Windows.cs#L145-L154

However, on Unix we just don't allow for paths that point to directories:
https://github.com/dotnet/runtime/blob/0696727ece4d363f171f80dd442b61fb20c6500a/src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.Unix.cs#L72

So in theory the following code could work on Unix so far:

```cs
File.WriteAllBytes($pathToApipe, new byte[1]);
```

Now it would throw. Would this be considered as a breaking change?

FWIW I would like to remove the Windows limitation (https://github.com/dotnet/runtime/pull/54676)

stephentoub replied:
> Now it would throw.

It shouldn't as of #55191.  We can add a test, though.

adamsitnik commented on src/libraries/System.Private.CoreLib/src/System/IO/File.cs:
```diff
@@ -373,19 +374,13 @@ public static void WriteAllBytes(string path, byte[] bytes)
+            fs.Write(bytes, 0, bytes.Length);
+#else
+            using SafeFileHandle sfh = OpenHandle(path, FileMode.Create, FileAccess.Write, FileShare.Read);
+            RandomAccess.Write(sfh, bytes, 0);
```
`Write` is the public method with a check for `CanSeek`:

https://github.com/dotnet/runtime/blob/ab64ce5f71479549222c99da3a15fd56f8328fe4/src/libraries/System.Private.CoreLib/src/System/IO/RandomAccess.cs#L146-L150

https://github.com/dotnet/runtime/blob/ab64ce5f71479549222c99da3a15fd56f8328fe4/src/libraries/System.Private.CoreLib/src/System/IO/RandomAccess.cs#L243-L251

to achieve the lack of throwing you need to call the internal `WriteAtOffset` that does not perform the check:

```suggestion
            RandomAccess.WriteAtOffset(sfh, bytes, 0);
```

stephentoub replied:
Maybe we need a NonRandomAccess ;-)

In all seriousness,  though,  maybe we should reconsider the type name,  and in this or a subsequent release add overloads that don't take a file offset, use the file's current position,  and thus don't require seeking.

adamsitnik replied:
Naming is hard ;)

> In all seriousness, though, maybe we should reconsider the type name in this release, and in this or a subsequent release add overloads that don't take a file offset, use the file's current position, and thus don't require seeking.

We could also allow for negative offsets that would mean "read next `n` bytes, ignore the offset". To be honest I don't have a perfect solution for it.

adamsitnik commented on src/libraries/System.Private.CoreLib/src/System/IO/File.cs:
```diff
@@ -836,22 +831,17 @@ public static Task WriteAllBytesAsync(string path, byte[] bytes, CancellationTok
 #else
-                await fs.WriteAsync(new ReadOnlyMemory<byte>(bytes), cancellationToken).ConfigureAwait(false);
+                using SafeFileHandle sfh = OpenHandle(path, FileMode.Create, FileAccess.Write, FileShare.Read, FileOptions.Asynchronous | FileOptions.SequentialScan);
+                await RandomAccess.WriteAsync(sfh, bytes, 0, cancellationToken).ConfigureAwait(false);
```
same as above:

```suggestion
                await RandomAccess.WriteAtOffsetAsync(sfh, bytes, 0, cancellationToken).ConfigureAwait(false);
```

adamsitnik commented on src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs:
```diff
@@ -20,10 +21,10 @@ public static void CopyFile(string sourceFullPath, string destFullPath, bool ove
             // Copy the contents of the file from the source to the destination, creating the destination in the process
-            using (var src = new FileStream(sourceFullPath, FileMode.Open, FileAccess.Read, FileShare.Read, DefaultBufferSize, FileOptions.None))
-            using (var dst = new FileStream(destFullPath, overwrite ? FileMode.Create : FileMode.CreateNew, FileAccess.ReadWrite, FileShare.None, DefaultBufferSize, FileOptions.None))
+            using (SafeFileHandle src = File.OpenHandle(sourceFullPath, FileMode.Open, FileAccess.Read, FileShare.Read, FileOptions.SequentialScan))
```
I wonder if `FileOptions.SequentialScan` is actually going to help as it requires an extra syscall on Unix:

https://github.com/dotnet/runtime/blob/d841f5e17897be5ef6aef5031b8a03ed3d8fd8dd/src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.Unix.cs#L278-L284

would be nice to benchmark it for small, medium and large files

stephentoub replied:
I'll remove it to unblock this PR.  But if we don't believe it's helpful on Unix, the right action isn't to avoid using it in cross-platform code but rather to stop issuing the PosixFAdvise call for the hint.

adamsitnik reviewed (approved):
LGTM, I hope that you enjoyed using the new APIs ;)

stephentoub committed:
Use RandomAccess/File.OpenHandle in a few more places

Places where we don't need to pay for the FileStream and the access pattern is easily replaced.
